### PR TITLE
feat: track campaign in rudderstack

### LINKED ISSUE #2003: Add UTM to data being sent to Rudderstack
**Proposal:**
We currently are capturing UTM somewhere as it is being sent to GA. We would like to capture as well in the event data being sent to Rudderstack through the Rudderstack JavaScript SDK

**Current behavior:**
Events being sent but does not include UTM

**Desired behavior:**
Same events being sent, but include UTM

**Alternatives considered:**
Very open to potential solutions, as long as we can get the data 

**Use case:**
For analytics and marketing

### PR DESCRIPTION
**Pull request by alespour**
Closes #2003 

This PR adds campaign data (UTM parameters) to events data send to Rudderstack.

### CODE REVIEW

**Comment by ui-ci-bot-influx** [comment]
Metric | Score | Level
------------ | ------------- | ------------- 
performance | 57 | :red_circle:
accessibility | 96 | :green_circle:
seo | 90 | :green_circle:
pwa | null | :hankey:
best-practices | 100 | :green_circle:


[Compare your scores with master branch](https://influxdb.aws.influxdata.io/orgs/844910ece80be8bc/dashboards/073c4b95f1089000?lower=now%28%29%20-%202d&vars%5Blighthouse_score_branch%5D=2263)

File: src/cloud/apis/reporting.ts [hunk 1]
```diff
@@ -51,3 +56,28 @@ export const reportPoints = (batch: PointsBatch) => {
     }
   }
 }
+
+const asRudderOptions = (tags: PointTags, campaign: object) => {
+  const options = {}
+  let context = null
+  Object.keys(tags).forEach((key: string) => {
+    const value = tags[key]
+    if (key != 'context') {
```
Comment (Asalem1) [inline src/cloud/apis/reporting.ts:65]:
```suggestion
    if (key !== 'context') {
```

File: src/cloud/apis/reporting.ts [hunk 2]
```diff
@@ -51,3 +56,28 @@ export const reportPoints = (batch: PointsBatch) => {
...
+      options[key] = value
+    } else {
+      if (typeof value != 'object') {
```
Comment (Asalem1) [inline src/cloud/apis/reporting.ts:68]:
```suggestion
      if (typeof value !== 'object') {
```

Also, do we care if the value is not an array?

File: src/cloud/apis/reporting.ts [hunk 3]
```diff
@@ -51,3 +56,28 @@ export const reportPoints = (batch: PointsBatch) => {
...
+        // context is reserved rudder element key and must be object, see rudder-sdk-js (#20463)
+        options['appContext'] = value
+      } else {
+        context = options['context'] = value
```
Comment (Asalem1) [inline src/cloud/apis/reporting.ts:72]: not really sure what this is about, can we separate this to make this a little clearer?

File: src/cloud/apis/reporting.ts [hunk 4]
```diff
@@ -51,3 +56,28 @@ export const reportPoints = (batch: PointsBatch) => {
...
+      }
+    }
+  })
+  if (Object.keys(campaign).length) {
+    if (context == null) {
```
Comment (Asalem1) [inline src/cloud/apis/reporting.ts:77]:
```suggestion
    if (context === null) {
```

File: src/cloud/apis/reporting.ts [hunk 5]
```diff
@@ -51,3 +56,28 @@ export const reportPoints = (batch: PointsBatch) => {
...
+      context = options['context'] = {}
```
Comment (Asalem1) [inline src/cloud/apis/reporting.ts:78]: again unclear on what this should do

File: src/cloud/utils/reporting.ts [hunk 6]
```diff
@@ -184,3 +188,26 @@ export const useLoadTimeReporting = (measurement: string) => {
...
+export const updateCampaignInfo = (query: string) => {
+  if (CLOUD && isFlagEnabled('rudderstackReporting')) {
+    const allowedUtmKeys = ['campaign', 'term', 'source', 'medium', 'content']
+    const queryParams = new URLSearchParams(query)
+    const info = {}
+    queryParams.forEach(function(value, key) {
+      if (key.startsWith('utm_')) {
+        let param = key.substr(4)
+        if (allowedUtmKeys.includes(param)) {
+          // see rudder-sdk-js (#2963)
```
Comment (Asalem1) [inline src/cloud/utils/reporting.ts:201]: is it possible to provide a direct link to the issue?

File: src/index.tsx [hunk 7]
```diff
@@ -39,6 +42,8 @@ updateReportingContext({
   session: cookieSession ? cookieSession[2].slice(5) : '',
 })
 
+updateCampaignInfo(window.location.search)
```
Comment (Asalem1) [inline src/index.tsx:45]: should we wrap this function call in a CLOUD condition?

File: src/cloud/apis/reporting.ts [hunk 8]
```diff
@@ -51,3 +56,28 @@ export const reportPoints = (batch: PointsBatch) => {
...
+    }
+    context['campaign'] = campaign
+  }
+  return options
+}
```
Comment (Asalem1) [inline src/cloud/apis/reporting.ts:83]: overall I'd love it if we can make this function a bit clearer. I understand that we're building out a rudderStackOption, but what does that look like? Is there a type we can reference that rudderstack's SDK provides? There are lots of nested conditionals that make this a little unclear what's happening where. Maybe that's a necessary evil, but maybe it doesn't have to be? We're also converting an object into an array in order to iterate over it, but why not just us a `for in` loop to iterate over the original object and deal with the data that way?

**Review by Asalem1 — CHANGES_REQUESTED** [review]

File: src/cloud/utils/reporting.ts (hunk 6 above)
Comment (alespour) [reply src/cloud/utils/reporting.ts]: Added link to source code (sorry using #N was misleading).

File: src/cloud/apis/reporting.ts (hunk 3 above)
Comment (alespour) [reply src/cloud/apis/reporting.ts]: I have refactored it, hopefully it is better now.

File: src/cloud/apis/reporting.ts (hunk 5 above)
Comment (alespour) [reply src/cloud/apis/reporting.ts]: I have refactored it, hopefully it is better now.

File: src/cloud/apis/reporting.ts (hunk 8 above)
Comment (alespour) [reply src/cloud/apis/reporting.ts:88]: I have refactored the whole function, hopefully it is better now. Unfortunately there is no reference in SDK that I could find. I had to use rudder SDK js source code and also their support on Slack channel(https://app.slack.com/client/T01E4PJL5J7/C01E4PLB135/thread/C01E4PLB135-1628077828.109800) 
image

**Review by Asalem1 — APPROVED** [review]
Not gonna hold this up on implementation details

**Pull request merged**